README.md
<div align="center">
<h1>Cryptorjs</h1>
Simple library for encryption and decryption of string, number and object using a key
<br/><br/>
<a href="https://travis-ci.org/fabioricali/Cryptor" target="_blank"><img src="https://travis-ci.org/fabioricali/Cryptor.svg?branch=master" title="Build Status"/></a>
<a href="https://coveralls.io/github/fabioricali/Cryptor?branch=master" target="_blank"><img src="https://coveralls.io/repos/github/fabioricali/Cryptor/badge.svg?branch=master" title="Coverage Status"/></a>
<a href="https://opensource.org/licenses/MIT" target="_blank"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" title="License: MIT"/></a>
<img src="https://img.shields.io/badge/team-terrons-orange.svg" title="Team Terrons"/>
</div>

## Installation

```javascript
npm install cryptorjs --save
```

## Example
### Basic
```javascript
var cryptorjs = require('cryptorjs');

var myCryptor = new cryptorjs('yourSecretKey');

var encoded = myCryptor.encode('myExampleString');
// => '37d8e07a3dddc2971f3e53b1021f51'

var decoded = myCryptor.decode('37d8e07a3dddc2971f3e53b1021f51');
// => 'myExampleString'
```
### Object encryption
```javascript
var cryptorjs = require('cryptorjs');

var myCryptor = new cryptorjs('yourSecretKey');

var encoded = myCryptor.encode({ a: 1, b: 2 });
// => '2183c42066819ed9184f1df116'

var decoded = myCryptor.decode('2183c42066819ed9184f1df116');
// => { a: 1, b: 2 }
```

### With a cipher
For example using "camellia-256-cfb1" cipher
```javascript
var cryptorjs = require('cryptorjs');

var myCryptor = new cryptorjs('yourSecretKey', 'camellia-256-cfb1');

var encoded = myCryptor.encode('myExampleString');
// => 'ac3277ba5c3f433d6b7ea70979fe55'

var decoded = myCryptor.decode('ac3277ba5c3f433d6b7ea70979fe55');
// => 'myExampleString'
```

### Ciphers
You can get the list with a static method
```javascript
var cryptorjs = require('cryptorjs');

cryptorjs.getCiphers();

/*=> [ 'aes-128-cbc',
        'aes-128-cbc-hmac-sha1',
        'aes-128-cbc-hmac-sha256',
        'aes-128-ccm',
        'aes-128-cfb',
        'aes-128-cfb1',
        'aes-128-cfb8',...]
        */
```

### Hash helpers
Create an hash using available hashes in your platform
```javascript
var cryptorjs = require('cryptorjs');

// Check if exists MD5 hash
console.log(cryptorjs.hasHash('md5')); //=> true

cryptorjs.hash('ciao', 'md5');

/*=> '6e6bc4e49dd477ebc98ef4046c067b5f'*/

cryptorjs.hash('ciao', 'sha1');

/*=> '1e4e888ac66f8dd41e00c5a7ac36a32a9950d271'*/
```

### Supported ciphers (since 3.0.0)
- aes-256-cbc
- aes-256-cbc-hmac-sha1
- aes-256-cbc-hmac-sha256
- aes-256-cfb
- aes-256-cfb1
- aes-256-cfb8
- aes-256-ctr
- aes-256-ofb
- aes256
- camellia-256-cbc
- camellia-256-cfb
- camellia-256-cfb1
- camellia-256-cfb8
- camellia-256-ofb
- camellia256

## License
Cryptorjs is open-sourced software licensed under the [MIT license](http://opensource.org/licenses/MIT)

## Author
[Fabio Ricali](http://rica.li)

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [3.0.0] - 2018-09-03
- Changed cipher method from `createCipher` to `createCipheriv` for node deprecation
- **Breaking changes** now support only this ciphers:
    - aes-256-cbc
    - aes-256-cbc-hmac-sha1
    - aes-256-cbc-hmac-sha256
    - aes-256-cfb
    - aes-256-cfb1
    - aes-256-cfb8
    - aes-256-ctr
    - aes-256-ofb
    - aes256
    - camellia-256-cbc
    - camellia-256-cfb
    - camellia-256-cfb1
    - camellia-256-cfb8
    - camellia-256-ofb
    - camellia256

## [2.1.0] - 2017-12-03
- Added static method `hasHash`
- Minor fix

## [2.0.2] - 2017-08-25
- Added static method `hash`
- Deprecated static methods `MD5` and `SHA1` 

## [1.2.0] - 2017-06-10
### Added
- Static methods `MD5` and `SHA1` 

## [1.1.0] - 2017-06-07
### Added
- CHANGELOG.md
- Support to encrypt object

### Fixed
- Some bug



test/main.js
/**
 * Created by Fabio on 21/05/2017.
 */
const assert = require('assert');
const cryptor = require('../index');
const helper = require('../src/helper');

describe('encode and decode', function () {

    it('should be equal', function () {
        let origin = 'myExampleString';
        let myCryptor = new cryptor('yourSecretKey');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);

        assert.equal(origin, decoded);
    });

    it('should be equal using a number', function () {
        let origin = 10000;
        let myCryptor = new cryptor('yourSecretKey');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);

        assert.equal(origin, decoded);
    });

    it('should be equal using a object', function () {
        let origin = {a:1,b:2};
        let myCryptor = new cryptor('yourSecretKey');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);
        console.log(origin, typeof origin);
        console.log(decoded, typeof decoded);

        assert.deepEqual(origin, decoded);
    });

    it('origin is empty', function () {
        let origin = '';
        let myCryptor = new cryptor('yourSecretKey');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);

        assert.equal(origin, decoded);
    });

    it('should be error if origin missing', function (done) {
        try {
            let myCryptor = new cryptor('yourSecretKey');
            myCryptor.encode();
            myCryptor.decode();
        }catch (error) {
            done();
        }
    });

    it('should be error if key missing', function (done) {
        try{
            new cryptor()
        }catch (error) {
            done();
        }
    });

    it('should be error if key is empty', function (done) {
        try{
            new cryptor('')
        }catch (error) {
            done();
        }
    });

    it('should be equal with "blowfish" cipher', function () {
        let origin = 'myExampleString';
        let myCryptor = new cryptor('yourSecretKey', 'blowfish');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);

        assert.equal(origin, decoded);
    });

});

describe('get algorithms', function () {
    it('should be array', function () {
        let algorithms = cryptor.getCiphers();
        console.log(algorithms);
        assert.equal(typeof algorithms, 'object');
    })
});

describe('helper function', function () {
    describe('normalizeInput', function () {
        it('should be return equal string', function () {
            let origin = 'hello';
            let result = helper.normalizeInput(origin);
            assert.equal(result, origin);
        });

        it('should be return equal json string', function () {
            let origin = {a:1};
            let result = helper.normalizeInput(origin);
            assert.equal(result, JSON.stringify(origin));
        });

        it('should be return equal number', function () {
            let origin = 100;
            let result = helper.normalizeInput(origin);
            assert.equal(result, origin);
            assert.strictEqual(Number(result), origin);
        });

        it('should be return error if null', function (done) {
            try {
                helper.normalizeInput(null);
            } catch (e) {
                done();
            }
        });
        it('should be return error if undefined', function (done) {
            try {
                helper.normalizeInput();
            } catch (e) {
                done();
            }
        });
    });

    describe('normalizeOutput', function () {
        it('should be return equal string', function () {
            let origin = 'test';
            let result = helper.normalizeOutput(origin);
            assert.equal(result, origin);
        });
        it('should be return equal object', function () {
            let origin = '{"a":1}';
            let result = helper.normalizeOutput(origin);
            assert.deepEqual(result, JSON.parse(origin));
        });
    });

    describe('hash helper', function () {
        it('should be return md5 hash', function () {
            let origin = 'ciao';
            let result = cryptor.md5(origin);
            console.log(origin, result);
            assert.equal(result, '6e6bc4e49dd477ebc98ef4046c067b5f');
        });
        it('should be return sha1 hash', function () {
            let origin = 'ciao';
            let result = cryptor.sha1(origin);
            console.log(origin, result);
            assert.equal(result, '1e4e888ac66f8dd41e00c5a7ac36a32a9950d271');
        });
        it('should be return sha512 hash', function () {
            let origin = 'ciao';
            let result = cryptor.sha512(origin);
            console.log(origin, result);
            assert.equal(result, 'a0c299b71a9e59d5ebb07917e70601a3570aa103e99a7bb65a58e780ec9077b1902d1dedb31b1457beda595fe4d71d779b6ca9cad476266cc07590e31d84b206');
        });
    });
});

test/helper.js
const helper = require('../src/helper');
const assert = require('assert');

describe('helper function', function () {
    describe('normalizeInput', function () {
        it('should be return equal string', function () {
            let origin = 'hello';
            let result = helper.normalizeInput(origin);
            assert.equal(result, origin);
        });

        it('should be return equal json string', function () {
            let origin = {a:1};
            let result = helper.normalizeInput(origin);
            assert.equal(result, JSON.stringify(origin));
        });

        it('should be return equal number', function () {
            let origin = 100;
            let result = helper.normalizeInput(origin);
            assert.equal(result, origin);
            assert.strictEqual(Number(result), origin);
        });

        it('should be return error if null', function (done) {
            try {
                helper.normalizeInput(null);
            } catch (e) {
                done();
            }
        });
        it('should be return error if undefined', function (done) {
            try {
                helper.normalizeInput();
            } catch (e) {
                done();
            }
        });
    });

    describe('normalizeOutput', function () {
        it('should be return equal string', function () {
            let origin = 'test';
            let result = helper.normalizeOutput(origin);
            assert.equal(result, origin);
        });
        it('should be return equal object', function () {
            let origin = '{"a":1}';
            let result = helper.normalizeOutput(origin);
            assert.deepEqual(result, JSON.parse(origin));
        });
    });
});

test/cryptor.js
/**
 * Created by Fabio on 21/05/2017.
 */
const assert = require('assert');
const cryptor = require('../index');
const be = require('bejs');

describe('encode and decode', function () {

    it('should be equal', function () {
        let origin = 'myExampleString';
        let myCryptor = new cryptor('yourSecretKey');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);

        assert.equal(origin, decoded);
    });

    it('should be equal using a number', function () {
        let origin = 10000;
        let myCryptor = new cryptor('yourSecretKey');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);

        assert.equal(origin, decoded);
    });

    it('should be equal using a object', function () {
        let origin = {a: 1, b: 2};
        let myCryptor = new cryptor('yourSecretKey');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);
        console.log(origin, typeof origin);
        console.log(decoded, typeof decoded);

        assert.deepEqual(origin, decoded);
    });

    it('origin is empty', function () {
        let origin = '';
        let myCryptor = new cryptor('yourSecretKey');
        let encoded = myCryptor.encode(origin);
        let decoded = myCryptor.decode(encoded);

        console.log(origin, encoded, decoded);

        assert.equal(origin, decoded);
    });

    it('should be error if origin missing', function (done) {
        try {
            let myCryptor = new cryptor('yourSecretKey');
            myCryptor.encode();
            myCryptor.decode();
        } catch (error) {
            done();
        }
    });

    it('should be error if key missing', function (done) {
        try {
            new cryptor()
        } catch (error) {
            done();
        }
    });

    it('should be error if key is empty', function (done) {
        try {
            new cryptor('')
        } catch (error) {
            done();
        }
    });

    it('should be equal with all cipher', function () {
        const origin = 'myExampleString';
        const okAlg = [];
        const algorithms = cryptor.getCiphers();
        algorithms.forEach(alg => {
            try {
                let myCryptor = new cryptor('yourSecretKey', alg);
                let encoded = myCryptor.encode(origin);
                let decoded = myCryptor.decode(encoded);
                assert.equal(origin, decoded);
                console.log('OK    -->', alg, encoded);
                okAlg.push(alg);
            } catch (e) {
                console.log('ERROR -->', alg, e.message)
            }
        });

        console.log(okAlg)
    });

    it('should be error, unsupported cipher', function (done) {
        try {
            new cryptor('yourSecretKey', 'aes-128-cbc');
        } catch (e) {
            console.log(e);
            done();
        }
    });

});

describe('get algorithms', function () {
    it('should be array', function () {
        let algorithms = cryptor.getCiphers();
        console.log(algorithms);
        assert.equal(typeof algorithms, 'object');
    })
});

describe('get hashes', function () {
    it('should be array', function () {
        let algorithms = cryptor.getHashes();
        console.log(algorithms);
        assert.equal(typeof algorithms, 'object');
    })
});

describe('hash', function () {
    it('should be return true', function () {
        let hash = cryptor.hash('hello', 'sha512');
        console.log(hash);
        be.err.sha512(hash);
    });
    it('should be return false', function () {
        let hash = cryptor.hash('hello', 'sha256');
        console.log(hash);
        be.err.not.sha512(hash);
    });
    it('should be return error', function (done) {
        try {
            cryptor.hash('hello', 'sha256444');
        } catch (e) {
            console.log(e.message);
            done()
        }
    });
});

describe('hasHash', function () {
    it('should be return true', function () {
        let result = cryptor.hasHash('sha512');
        console.log(result);
        be.err.true(result);
    });
    it('should be return false', function () {
        let result = cryptor.hasHash('foooo');
        console.log(result);
        be.err.false(result);
    });
});

describe('hash helper', function () {
    it('should be return md5 hash', function () {
        let origin = 'ciao';
        let result = cryptor.md5(origin);
        console.log(origin, result);
        assert.equal(result, '6e6bc4e49dd477ebc98ef4046c067b5f');
    });
    it('should be return sha1 hash', function () {
        let origin = 'ciao';
        let result = cryptor.sha1(origin);
        console.log(origin, result);
        assert.equal(result, '1e4e888ac66f8dd41e00c5a7ac36a32a9950d271');
    });
});


src/cryptor.js
/**
 * Created by Fabio on 21/05/2017.
 */
'use strict';
const crypto = require('crypto');
const helper = require('./helper');
const deprecate = require('depreca');
const {ALGORITHM} = require('./constants');
/**
 * Cryptor class
 */
class Cryptor {

    /**
     * Cryptor constructor
     * @param key
     * @param algorithm
     */
    constructor(key, algorithm = 'aes-256-ctr') {
        if (typeof key !== 'string')
            throw new Error('required an string key');

        if (key === '')
            throw new Error('key cannot be empty');

        if (!ALGORITHM.includes(algorithm))
            throw new Error(`algorithm ${algorithm} not supported, use those available: ${ALGORITHM.join(', ')}`);

        // Transform to 32 chars
        key = this.constructor.hash(key, 'md5');

        Object.defineProperties(this, {
            algorithm: {
                value: algorithm
            },
            key: {
                value: key
            },
            iv: {
                value: key.substr(16)
            },
            options: {
                value: {}
            }
        });
    }

    /**
     * Encode string
     * @param str
     * @return {string}
     */
    encode(str) {
        str = helper.normalizeInput(str);
        const cipher = crypto.createCipheriv(this.algorithm, this.key, this.iv, this.options);
        return cipher.update(str, 'utf8', 'hex') + cipher.final('hex');
    }

    /**
     * Decode string
     * @param str
     * @return {string}
     */
    decode(str) {
        str = helper.normalizeInput(str);
        const decipher = crypto.createDecipheriv(this.algorithm, this.key, this.iv, this.options);
        const decoded = decipher.update(str, 'hex', 'utf8') + decipher.final('utf8');
        return helper.normalizeOutput(decoded);
    }

    /**
     * Get available ciphers
     * @return {array}
     */
    static getCiphers() {
        return crypto.getCiphers();
    }

    /**
     * Get available hashes
     * @return {array}
     */
    static getHashes() {
        return crypto.getHashes();
    }

    /**
     * MD5 hash
     * @param str
     * @returns {*}
     * @deprecated
     */
    static md5(str) {
        deprecate('md5 is deprecated, use hash method instead. e.g. hash("your string", "md5")');
        return crypto.createHash('md5').update(str).digest('hex');
    }

    /**
     * SHA1 hash
     * @param str
     * @returns {*}
     * @deprecated
     */
    static sha1(str) {
        deprecate('sha1 is deprecated, use hash method instead. e.g. hash("your string", "sha1")');
        return crypto.createHash('sha1').update(str).digest('hex');
    }

    /**
     * Creates hash of an string based on available hashes of platform
     * @param str
     * @param hash
     * @returns {*}
     */
    static hash(str, hash) {
        if (Cryptor.hasHash(hash)) {
            return crypto.createHash(hash).update(str).digest('hex');
        } else {
            throw new Error('hash ' + hash + ' not found in your platform')
        }
    }

    /**
     * Check if hash exists
     * @param hash
     * @returns {boolean}
     */
    static hasHash(hash) {
        return Cryptor.getHashes().indexOf(hash) !== -1;
    }
}

module.exports = Cryptor;

